File: ParkingApp/App/ms-audit/src/common/cache.service.ts
import { Inject, Injectable, Logger } from '@nestjs/common';
import { CACHE_MANAGER } from '@nestjs/cache-manager';
import type { Cache } from 'cache-manager';

@Injectable()
export class CacheService {
  private readonly logger = new Logger(CacheService.name);

  constructor(
    @Inject(CACHE_MANAGER)
    private readonly cacheManager: Cache,
  ) {}

  async get<T>(key: string): Promise<T | null> {
    try {
      const value = await this.cacheManager.get<T>(key);

      if (value === undefined || value === null) {
        this.logger.log(`Cache get: ${key} -> MISS`);
        return null;
      }

      this.logger.log(`Cache get: ${key} -> HIT`);

      return value;
    } catch (error) {
      this.logger.error(
        `Error obteniendo la clave ${key}: ${
          error instanceof Error
            ? error.message
            : String(error)
        }`,
      );

      return null;
    }
  }

  async set<T>(
    key: string,
    value: T,
    ttlMilliseconds = 300000,
  ): Promise<void> {
    try {
      await this.cacheManager.set(
        key,
        value,
        ttlMilliseconds,
      );

      this.logger.log(
        `Cache set: ${key} (TTL: ${
          ttlMilliseconds / 1000
        }s)`,
      );
    } catch (error) {
      this.logger.error(
        `Error guardando la clave ${key}: ${
          error instanceof Error
            ? error.message
            : String(error)
        }`,
      );
    }
  }

  async del(key: string): Promise<void> {
    try {
      await this.cacheManager.del(key);

      this.logger.log(`Cache del: ${key}`);
    } catch (error) {
      this.logger.error(
        `Error eliminando la clave ${key}: ${
          error instanceof Error
            ? error.message
            : String(error)
        }`,
      );
    }
  }
}

File: ParkingApp/App/ms-usuarios-roles-oauth/src/main/java/ec/edu/espe/usuarios/cache/CacheService.java
package ec.edu.espe.usuarios.cache;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Service;

import java.time.Duration;

@Slf4j
@Service
@RequiredArgsConstructor
public class CacheService {

    private final StringRedisTemplate redisTemplate;
    private final ObjectMapper objectMapper;

    public <T> T get(
            String key,
            Class<T> type
    ) {
        try {
            String json = redisTemplate
                    .opsForValue()
                    .get(key);

            if (json == null) {
                log.info(
                        "Cache get: {} -> MISS",
                        key
                );

                return null;
            }

            log.info(
                    "Cache get: {} -> HIT",
                    key
            );

            return objectMapper.readValue(
                    json,
                    type
            );

        } catch (Exception exception) {
            log.error(
                    "Error obteniendo la clave {}: {}",
                    key,
                    exception.getMessage()
            );

            return null;
        }
    }

    public <T> T get(
            String key,
            TypeReference<T> type
    ) {
        try {
            String json = redisTemplate
                    .opsForValue()
                    .get(key);

            if (json == null) {
                log.info(
                        "Cache get: {} -> MISS",
                        key
                );

                return null;
            }

            log.info(
                    "Cache get: {} -> HIT",
                    key
            );

            return objectMapper.readValue(
                    json,
                    type
            );

        } catch (Exception exception) {
            log.error(
                    "Error obteniendo la clave {}: {}",
                    key,
                    exception.getMessage()
            );

            return null;
        }
    }

    public void set(
            String key,
            Object value,
            long ttlSeconds
    ) {
        try {
            String json = objectMapper
                    .writeValueAsString(value);

            redisTemplate
                    .opsForValue()
                    .set(
                            key,
                            json,
                            Duration.ofSeconds(
                                    ttlSeconds
                            )
                    );

            log.info(
                    "Cache set: {} (TTL: {}s)",
                    key,
                    ttlSeconds
            );

        } catch (Exception exception) {
            log.error(
                    "Error guardando la clave {}: {}",
                    key,
                    exception.getMessage()
            );
        }
    }

    public void delete(String key) {
        try {
            redisTemplate.delete(key);

            log.info(
                    "Cache del: {}",
                    key
            );

        } catch (Exception exception) {
            log.error(
                    "Error eliminando la clave {}: {}",
                    key,
                    exception.getMessage()
            );
        }
    }

    public boolean exists(String key) {
        try {
            Boolean exists =
                    redisTemplate.hasKey(key);

            boolean result =
                    Boolean.TRUE.equals(exists);

            log.info(
                    "Cache exists: {} -> {}",
                    key,
                    result ? "HIT" : "MISS"
            );

            return result;

        } catch (Exception exception) {
            log.error(
                    "Error comprobando la clave {}: {}",
                    key,
                    exception.getMessage()
            );

            return false;
        }
    }
}